### DISCUSSION TITLE
Combining multiple mp3 files to be returned as a single MediaStreamTrack

### DISCUSSION THREAD

**Discussion by pushkarprasad007** [aiortc/aiortc#1076] [Q&A]
I will be using LLM (like GPT) to generate an answer - which would then be converted to speech, which I want to send over to the browser using aiortc. However, since LLM take time to produce complete output, instead of waiting for it to complete, we can read partial answers as soon it appears, and every few words, generate mp3 file for those many words, and then stream those. So not all the mp3 files would be available immediately, and instead, I need to keep on adding them as soon they appear (say every 4-5 words) from LLM.

I wrote a custom MediaStreamTrack to achieve the same. I have tried this with 2 files, a.mp3 and b.mp3. 

I ran across 2 issues:
* The last few hundred millisecond of a.mp3 sound stretched. 
* The initial 2 (or so) second of b.mp3 go blank, and then after that it plays

Clearly, the addition of frames need to be done better so that this can work. I am definitely missing something here - would be great if someone can point me in the right direction. 

```
class CombinedAudioTrack(MediaStreamTrack):
    """
    An audio track which reads from multiple mp3
    """
    kind = "audio"
    currentMediaPlayer:MediaStreamTrack = None
    queue = asyncio.Queue()
    _stop: float = False

    def __init__(self) -> None:
        super().__init__()
        # self.readyState = "live"

    def addNewMP3File(self, mp3File, last:bool = False):
        self.queue.put_nowait(mp3File)
        if last:
            self._stop = True
    
    async def getNextMediaStreamTrack(self):
        mp3File = await self.queue.get()
        self.currentMediaPlayer = MediaPlayer(os.path.join(ROOT, mp3File)).audio

    async def recv(self) -> Frame:
        print("Came in recv")
        try:
            # Should only happen first time
            if not self.currentMediaPlayer:
                await self.getNextMediaStreamTrack()    
            frame = await self.currentMediaPlayer.recv()
            print(frame)
            return frame
        except MediaStreamError:
            # Its time to move the current media player forward
            if(self._stop):
                # self.stop()
                raise MediaStreamError()       
            await self.getNextMediaStreamTrack()
            return await self.currentMediaPlayer.recv()
```

**Comment by pushkarp-sharp** [comment]
Any update?

**Comment by Antonyesk601** [comment]
Hey I have done something a bit similar and i suggest looking at using some form of getting you tts output in some stream friendly format. I went the easy route of getting all my audio as pcm16 16000. Makes life a lot easier. the downside is you might probably need to manage some stuff yourself like timestamps, frame chunking etc

As for your MediaStreamTrack, I think you might be running into an issue with the timestamps of the frames as it feels like the timestamps would reset on each new audio file you're reading which might confuse the recipient

**Reply by pushkarp-sharp** [reply]
Can you suggest a code for this please?

**Comment by lalanikarim** [comment]
Hi @pushkarprasad007 
I am working on a similar project involving LLMs and ran into similar issue. 
Take a look at this implementation of a custom MediaStreamTrack https://github.com/lalanikarim/webrtc-ai-voice-chat/blob/main/playback_stream_track.py 
I start with a one second silence wav file, which I re-queue until I have audio from my text to audio model. At that point I queue the audio from the generated audio. I then re-queue the silence track until I have the next piece of audio generated and ready to play.
Like  @Antonyesk601 suggested, AudioFrames from new tracks always start at time 0. For instance, if I queue my audio file to play soon after the one second of silence wav file, I will not hear the first one second of audio from my generated file, since the MediaPlayer has already moved one second past the start of the audio.
You will need to increase the `AudioFrame.pts` for any new frames you append enough to bump the generated time value to the current time step.
I hope this help.

**Reply by pushkarprasad007** [reply]
@lalanikarim Can you tell me if I want to store the recording in a mp3, what are the changes I'd need to do? Thanks for your super awesome repo!

**Reply by lalanikarim** [reply]
> @lalanikarim Can you tell me if I want to store the recording in a mp3, what are the changes I'd need to do? Thanks for your super awesome repo!

It will probably be in this section 
https://github.com/lalanikarim/webrtc-ai-voice-chat/blob/main/server.py#L116-L118

Here you will have the audio data from the recorder. You should be able to convert it into an MP3 file. 

If I come across a package that lets you do that, I'll share it here.

**Reply by pushkarp-sharp** [reply]
That would be superb - thank you @lalanikarim !
